sql/core/src/main/scala/org/apache/spark/sql/expressions/udaf.scala

```diff
@@ -165,3 +165,140 @@ abstract class MutableAggregationBuffer extends Row {
   /** Update the ith value of this buffer. */
   def update(i: Int, value: Any): Unit
 }
+
+/**
+ * The base class for implementing user-defined imperative aggregator (UDIA).
+ *
+ * @tparam A the aggregating type: implements the aggregation logic.
+ *
+ * @since 3.0.0
+ */
+@Experimental
+abstract class UserDefinedImperativeAggregator[A] extends Serializable {
```

Comment:
This looks very similar to the existing API `Aggregator`. Have you tried it before?

Reply:
See some examples in https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala

Reply:
Internally it's executed by either `SimpleTypedAggregateExpression` or `ComplexTypedAggregateExpression`.
* `SimpleTypedAggregateExpression`: This will be picked if the agg buffer is a single primitive type. In this case, serde is very fast and it's OK to do it for each input.
* `ComplexTypedAggregateExpression`: This only do serde once until the buffer needs to be sent out through shuffle.

Reply:
There are some definite similarities, but Aggregator is designed specifically for use on scala DataSet, and I was going for something that operates on DataFrame, closer to UDAF.

For example, UserDefinedImperativeAggregator can make use of UserDefinedType (like UDAF), which will allow it to be exposed via pyspark like UDAF does. I don't think Aggregator can do that.

Reply:
DataFrame is `Dataset[Row]`, so `Aggregator[Row, BUF, OUT]` can operate on DataFrame AFAIK.

It seems we can also add a new registering API that can register an `Aggregator[Row, BUF, OUT]` as an UDAF and call it in SQL/PySpark.

cc @viirya to confirm.

Reply:
Certainly agreed wrt `DataSet[Row]` as far as that goes.

There are some other boxes to check. Off the top of my head:
* Can an `Encoder` be built around `UserDefinedType`?  (maybe)  If so will pyspark be able to use it in the same way?
* `Aggregator` has `toColumn`, which returns a `TypedColumn`, compared to UDAF, which has `apply` and `distinct`, returning `Column` - what are the implications of those differences?
* as you say, `register`.  Getting that to work with UDIA was pretty easy, so I'd guess that is feasible.

To me, it is very important that a solution operates via pyspark in the way that UDAF does. 

I would also prefer something that "can pass all the same unit tests as UDAF", which I believe I was able to do with UDIA. I don't think `Aggregator` exactly does that, although it might be reasonably close for most purposes.

Reply:
After exploring what an enhancement to `Aggregator` looks like, my opinion is still that UDIA is a better replacement for UDAF.  Applying `Aggregator` in this way feels like hammering a square peg into a round hole.